### PR TITLE
Replace curl with wget; see #1567

### PR DESCRIPTION
**Pull request by callahad**
Let's see if this passes CI...

(it should, but there are some subtleties in curl vs wget that I'm going to double-check while CI runs)

### CODE REVIEW

**Comment by josegonzalez** [comment]
Nice! I'm in the middle of making a release, but this will go into 0.4.5 :)

**Comment by callahad** [comment]
Heh. You're presuming it won't Break Everything. Waiting for 0.4.5 is a good idea :)

File: docs/troubleshooting.md [hunk 1]
```diff
@@ -138,12 +138,12 @@ If you see output similar this when deploying:
  !     tar: Error is not recoverable: exiting now
 ```
 
-it might that the curl command that is supposed to fetch the buildpack (anything in the low megabyte file size range) takes too long to finish, due to slowish connection.  To overwrite the default values (connection timeout: 3 seconds, total maximum time for operation: 30 seconds), edit `/home/dokku/ENV` like the following:
+it might that the command that is supposed to fetch the buildpack (anything in the low megabyte file size range) takes too long to finish, due to slowish connection.  To overwrite the default values (connection timeout: 3 seconds, total maximum time for operation: 30 seconds), edit `/home/dokku/ENV` like the following:
```
Comment (josegonzalez) [inline docs/troubleshooting.md]: The english here was always solid :P

File: plugins/build-env/pre-build-buildpack [hunk 2]
```diff
@@ -6,8 +6,8 @@ source "$PLUGIN_AVAILABLE_PATH/config/functions"
 APP="$1"; IMAGE=$(get_app_image_name $APP); BUILD_ENV=""
 verify_app_name "$APP"
 
-[[ -z $(config_get --global CURL_CONNECT_TIMEOUT) ]] && config_set --global CURL_CONNECT_TIMEOUT=5
-[[ -z $(config_get --global CURL_TIMEOUT) ]] && config_set --global CURL_TIMEOUT=30
+[[ -z $(config_get --global WGET_CONNECT_TIMEOUT) ]] && config_set --global WGET_CONNECT_TIMEOUT=5
+[[ -z $(config_get --global WGET_TIMEOUT) ]] && config_set --global WGET_TIMEOUT=30
```
Comment (josegonzalez) [inline plugins/build-env/pre-build-buildpack]: These are internal environment variables for the buildpacks, so they should be reverted.

**Comment by callahad** [comment]
So, one potential issue with this is that both `wget --quiet` and `curl --silent` are both completely silent, and we still want to see log entries when something fails.

Curl has `--show-error`, to still print when something blows up, but wget doesn't have anything like it. The closest wget has is `--no-verbose`, which prints a single status line for all requests, in addition to any errors:

```
# curl -sS is silent, but still prints errors
$ curl -sSL 'http://httpbin.org/status/200'
$ curl -sSL 'http://httpbin.org/status/404'
$ curl -sSL 'http://httpbin.org/status/500'
$ curl -sSL 'http://example.invalid'
curl: (6) Could not resolve host: example.invalid

# wget -q is completely silent
$ wget -qO- 'http://httpbin.org/status/200'
$ wget -qO- 'http://httpbin.org/status/404'
$ wget -qO- 'http://httpbin.org/status/500'
$ wget -qO- 'http://example.invalid'
$ wget -qO- 'http://example.invalid'

# wget -nv is almost silent, but prints on every request
$ wget -nv -O- 'http://httpbin.org/status/200'
2015-11-06 08:17:57 URL:http://httpbin.org/status/200 [0/0] -> "-" [1]
$ wget -nv -O- 'http://httpbin.org/status/400'
http://httpbin.org/status/404:
2015-11-06 08:18:00 ERROR 404: NOT FOUND.
$ wget -nv -O- 'http://httpbin.org/status/500'
http://httpbin.org/status/500:
2015-11-06 08:18:05 ERROR 500: INTERNAL SERVER ERROR.
$ wget -nv -O- 'http://example.invalid'
wget: unable to resolve host address 'example.invalid'
```

We can either:

- lose printing on errors with `wget -q`
- get a print on every request with `wget -nv`
- try to emulate curl's behavior with `wget -nv ... &2>1 | grep '^wget: '`

**Comment by callahad** [comment]
And, if we need to be able to pass the output along, we can send stderr into a subshell, and stdout wherever

```bash
$ wget -nv -O - 'http://example.invalid' 2> >(grep '^wget: ') > /dev/null
wget: unable to resolve host address 'example.invalid'
```

**Comment by josegonzalez** [comment]
I do a trick where I redirect all output to a temporary file, and if there is an error, I dump the contents of that file. See [here](https://github.com/seatgeek/bash-aptfile/blob/master/bin/aptfile#L130) and [here](https://github.com/seatgeek/bash-aptfile/blob/master/bin/aptfile#L115).

**Comment by callahad** [comment]
Yeah, I need to revisit things, especially around what output goes where in the bats tests.

**Comment by callahad** [comment]
Dropping this PR. We *could* probably make wget to work everywhere, but curl is *way* more developer friendly for us, especially its `--show-error` flag. Instead of excising it completely, let's just avoid using curl during bootstrapping, and keep the .deb package dependency on curl, so that we can use it once Dokku is installed.

**Pull request merged**